mono/sgen/sgen-gc.c

```diff
@@ -2095,10 +2090,11 @@ major_start_concurrent_collection (const char *reason)
 
 	binary_protocol_concurrent_start ();
 
+	init_gray_queue (&gc_thread_gray_queue, TRUE);
 	// FIXME: store reason and pass it when finishing
-	major_start_collection (TRUE, NULL);
-
+	major_start_collection (&gc_thread_gray_queue, TRUE, NULL);
 	sgen_workers_take_from_queue_and_awake (&gc_thread_gray_queue);
+	sgen_gray_object_queue_dispose (&gc_thread_gray_queue);
 
```

Comment:
Given we properly dispose the queue here, we should no longer free the sections in finish_gray_stack. As a side note, it also seems that we don't free the sections of the private_gray_queues of the workers.